### PR TITLE
feat(core/ui): trap focus inside modals

### PR DESCRIPTION
**Pull request by SuyashSalampuria**

### CODE REVIEW

**Comment by SuyashSalampuria** [comment]
The only problem I am facing now is the security issue
```
Permission denied to access property "document" on cross-origin object
```
when trying to get the contentDocument of Iframe. I will get through it by tommorow

**Comment by SuyashSalampuria** [comment]
@saschanaz @sidvishnoi I researched a lot but came to know that Cross site contents can not be read by javascript. No major browser will allow you that.So, it won't be possible to trap focus for an iframe

**Comment by saschanaz** [comment]
We might want to add invisible focusable elements to allow trapping.

**Comment by sidvishnoi** [comment]
Button added by https://github.com/w3c/respec/pull/2697 will be focussable, right?

**Comment by saschanaz** [comment]
> Button added by #2697 will be focusable, right?

Yes but it won't be sufficient as we also need an element *after* iframe.

**Comment by SuyashSalampuria** [comment]
Shall I add something like 
<button style="background:transparent; border:none; color:transparent;">invisible</button>
at the end of the modal.
The problem with this way is that the focus will be lost according to the user when it will be focussing on this button.
Do you have a better way @sidvishnoi  @saschanaz

**Comment by sidvishnoi** [comment]
> the focus will be lost according to the user

Correct. 

Can we add `iframe` to `focusableEls` in `trapFocus(element)`? `iframe` is focusable, so we should be able to cycle focus without adding invisible element. Can you try?

**Comment by sidvishnoi** [comment]
...that might not work. One way could be: after tabbing, if `document.activeElement` is our invisible element, immediately call `firstFocusableEl.focus()`

**Comment by SuyashSalampuria** [comment]
> ...that might not work. One way could be: after tabbing, if `document.activeElement` is our invisible element, immediately call `firstFocusableEl.focus()`

Any children of elements with display:none  are not in the focus order.
So, we need to check using different styles actually

**Comment by SuyashSalampuria** [comment]
Although ,I have pushed it, But I dont find satisfactory behaviour at all. Please suggest if you find better ways!!

**Comment by saschanaz** [comment]
IMO we don't need any special handling for iframe (as it won't work). Maybe insert a footer, e.g. "File an issue" button?

**Comment by SuyashSalampuria** [comment]
Is this kind of behaviour acceptable ? I will refactor the code well

**Comment by saschanaz** [comment]
LGTM for the behavior. I think we should insert the additional focusable item only in the definition search, though.

File: src/core/ui.js [hunk 1]
```diff
@@ -234,6 +236,17 @@ export const ui = {
     overlay.addEventListener("click", () => this.closeModal(currentOwner));
     overlay.classList.toggle("respec-show-overlay");
     modal.hidden = false;
+    if (content.tagName === "IFRAME") {
+      modal.insertAdjacentHTML(
```
Comment (sidvishnoi) [inline src/core/ui.js]: I've an idea - add a footer on iframe one.: a link that asks: "open this UI in new tab"
WDYT @saschanaz ?

Comment (saschanaz) [reply src/core/ui.js]: That sounds good.

Comment (saschanaz) [reply src/core/ui.js]: BTW, I'd prefer to insert the whatever footer in ui/search-xref rather than core/ui.

Comment (sidvishnoi) [reply src/core/ui.js]: @SuyashSalampuria 
Added a link after iframe with https://github.com/w3c/respec/pull/2712, no need to add invisible button anymore.

**Comment by SuyashSalampuria** [comment]
Closes #884

File: src/core/ui.js [hunk 2]
```diff
@@ -96,7 +96,9 @@ function trapFocus(element) {
   );
   const firstFocusableEl = focusableEls[0];
   const lastFocusableEl = focusableEls[focusableEls.length - 1];
-
+  if (focusableEls) {
```
Comment (sidvishnoi) [inline src/core/ui.js]: `focusableEls` is always truthy
```suggestion
  if (firstFocusableEl) {
```

**Comment by sidvishnoi** [comment]
Focus is not getting trapped in `About ReSpec` popup. Can you confirm?
Overall, this is really awesome! Great work @SuyashSalampuria :100:

**Comment by SuyashSalampuria** [comment]
> Focus is not getting trapped in `About ReSpec` popup. Can you confirm?
> Overall, this is really awesome! Great work @SuyashSalampuria 100

Indeed it is not . Let  me check why?

**Comment by sidvishnoi** [comment]
I just checked. When `trapFocus` is called, the modal only has the close button. It takes a few moments to render rest of it, so trapping doesn't work.

**Comment by sidvishnoi** [comment]
Can you try moving https://github.com/w3c/respec/blob/4b6ef28013c2caaecbd856143282cde1840596e8/src/ui/about-respec.js#L36 to be after the "render`" call?

**Comment by SuyashSalampuria** [comment]
> Can you try moving
> 
> https://github.com/w3c/respec/blob/4b6ef28013c2caaecbd856143282cde1840596e8/src/ui/about-respec.js#L36
> to be after the "render`" call?

It worked absolutely!!

File: src/ui/about-respec.js [hunk 3]
```diff
@@ -33,7 +33,7 @@ const button = ui.addCommand(
 );
 
 function show() {
-  ui.freshModal(`${l10n.about_respec} - ${window.respecVersion}`, div, button);
+  // ui.freshModal(`${l10n.about_respec} - ${window.respecVersion}`, div, button);
```
Comment (sidvishnoi) [inline src/ui/about-respec.js]:
```suggestion
```

**Comment by SuyashSalampuria** [comment]
@saschanaz @sidvishnoi I see that the time taken for loading the respec fill is nearly 3 seconds, which is too high. Any idea on how we can decrease it

**Comment by saschanaz** [comment]
>I see that the time taken for loading the respec fill is nearly 3 seconds, which is too high. Any idea on how we can decrease it

You might want to click the About menu item and see what modules take too much time. Maybe there can be some lines to optimize.

Alternatively, we might want to run some modules in parallel, as currently all modules run in serial way.

**Pull request merged**